## import 모듈

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots
from warnings import filterwarnings
filterwarnings('ignore')

colors = ['#B1EDED','#B1B2ED','#1DE7ED','#1DA5ED','#1D50ED','#16548E']
gen_colors = ['#4169E1','#B2182B','#81007F','#D1B2FF','#EFE4E2']
jp_colors = ['#D90B0B','#F24444','#EFE4E2','#FCCE88','#64807F']
cn_colors = ['#E0201B','#FFCE3F','#A63F03','#04BF33','#F2E6D8']

## import data

In [2]:
df21 = pd.read_csv("https://raw.githubusercontent.com/hangack/project-green/main/Kaggle_Survey-2021/data/kaggle-survey-2021/kaggle_survey_2021_responses.csv", dtype='unicode')
df19 = pd.read_csv("https://raw.githubusercontent.com/hangack/project-green/main/Kaggle_Survey-2021/data/kaggle-survey-2019/multiple_choice_responses.csv", dtype='unicode')

#df21 = pd.read_csv("D:/_Bdata/Codding-base-Python/jupyter_nb/Kaggle Survey - 2021 Analysis - Plotly/kaggle_survey_2021_responses.csv", dtype='unicode')
#df19 = pd.read_csv("D:/_Bdata/Codding-base-Python/jupyter_nb/Kaggle Survey - 2021 Analysis - Plotly/multiple_choice_responses.csv", dtype='unicode')

#df21 = pd.read_csv("E:/Fear/Univ/Big_data/Training/Github/Codding-base-Python/jupyter_nb/Kaggle Survey - 2021 Analysis - Plotly/kaggle_survey_2021_responses.csv", dtype='unicode')
#df19 = pd.read_csv("E:/Fear/Univ/Big_data/Training/Github/Codding-base-Python/jupyter_nb/Kaggle Survey - 2021 Analysis - Plotly/multiple_choice_responses.csv", dtype='unicode')

#df21 = pd.read_csv("../input/kaggle-survey-2021/kaggle_survey_2021_responses.csv", dtype='unicode')
#df19 = pd.read_csv("../input/kaggle-survey-2019/multiple_choice_responses.csv", dtype='unicode')

## Japan & China: Programming_Languages

### 2021 Japan & China total
2021년 Q3(Country) 일본 중국 추출 dataframe

In [3]:
df21_ChJp = df21[df21.Q3.isin(["Japan","China"])]

### 2019 Japan & China total
2021년 Q3(Country) 일본 중국 추출 dataframe

In [4]:
df19_ChJp = df19[df19.Q3.isin(["Japan","China"])]

### Split Country

나라별 value_counts를 위해 각 나라로 dataframe 분리

#### 2021

In [5]:
df21_Ch = df21_ChJp[df21_ChJp.Q3.isin(["China"])]
df21_Jp = df21_ChJp[df21_ChJp.Q3.isin(["Japan"])]


## Q7(Program_Language): 칼럼번호 8~20 - others
df21_Jp_PL = pd.DataFrame()
df21_Jp_PL['Program_Language'] = [df21_Jp[col][1:].value_counts().index[0] for col in df21_Jp.columns[7:20]]
df21_Jp_PL['counts'] = [df21_Jp[col][1:].value_counts().values[0] for col in df21_Jp.columns[7:20]]


## Q7(Program_Language): 칼럼번호 8~20 - others
df21_Ch_PL = pd.DataFrame()
df21_Ch_PL['Program_Language'] = [df21_Ch[col][1:].value_counts() .index[0] for col in df21_Ch.columns[7:20]]
df21_Ch_PL['counts'] = [df21_Ch[col][1:].value_counts() .values[0] for col in df21_Ch.columns[7:20]]



## 제거된 나라 칼럼과 value를 각각 삽입 및 통합
df21_Jp_PL.insert(0, 'Country',  'Japan')
df21_Ch_PL.insert(0, 'Country',  'China')

df21_PL_JnC = pd.concat([df21_Jp_PL,df21_Ch_PL], ignore_index=True)

IndexError: index 0 is out of bounds for axis 0 with size 0

##### 결측 column 식별 및 제거

***IndexError**: index 0 is out of bounds for axis 0 with size 0
오류가 식별됐다.
아마 China, Program_Language의 특정 응답이 없어서 발생한거같다.

N/A 개수를 식별해보자

In [8]:
print("df21_Ch\'s rows_num:",len(df21_Ch))
print(df21_Ch.isnull().sum().iloc[7:20])

df21_Ch's rows_num: 814
Q7_Part_1      76
Q7_Part_2     729
Q7_Part_3     599
Q7_Part_4     588
Q7_Part_5     546
Q7_Part_6     602
Q7_Part_7     728
Q7_Part_8     810
Q7_Part_9     809
Q7_Part_10    783
Q7_Part_11    645
Q7_Part_12    814
Q7_OTHER      787
dtype: int64


찾았다. Q7_Part_12 이 녀석이 문제였다.
 - 행 개수는 814개고 Part_12의 N/A 개수도 814개이다.

#####

위 식별 과정의 시행 횟수가 많아진다면 if 문을 사용한 define을 사용해도 되겠지만 일단은 직접 제거하겠다.

In [ ]:
df21_Ch = df21_ChJp[df21_ChJp.Q3.isin(["China"])]
df21_Jp = df21_ChJp[df21_ChJp.Q3.isin(["Japan"])]


## Q7(Program_Language): 칼럼번호 8~20 - others
df21_Jp_PL = pd.DataFrame()
df21_Jp_PL['Program_Language'] = [df21_Jp[col][1:].value_counts().index[0] for col in df21_Jp.columns[7:19]]
df21_Jp_PL['counts'] = [df21_Jp[col][1:].value_counts().values[0] for col in df21_Jp.columns[7:19]]


## 2021 China: Q7_Part12(None) value == 0이므로 결측값 제거
df21_Ch_rmQ07P12 = df21_Ch.drop(['Q7_Part_12'], axis='columns')

## Q7(Program_Language): 칼럼번호 8~20 - others - Q7_Part12(None)
df21_Ch_PL = pd.DataFrame()
df21_Ch_PL['Program_Language'] = [df21_Ch_rmQ07P12[col][1:].value_counts() .index[0] for col in df21_Ch_rmQ07P12.columns[7:18]]
df21_Ch_PL['counts'] = [df21_Ch_rmQ07P12[col][1:].value_counts() .values[0] for col in df21_Ch_rmQ07P12.columns[7:18]]


## 제거된 나라 칼럼과 value를 각각 삽입 및 통합
df21_Jp_PL.insert(0, 'Country',  'Japan')
df21_Ch_PL.insert(0, 'Country',  'China')

df21_PL_JnC = pd.concat([df21_Jp_PL,df21_Ch_PL], ignore_index=True)

#### 2019
위와 동일 과정

In [ ]:
df19_Ch = df19_ChJp[df19_ChJp.Q3.isin(["China"])]
df19_Jp = df19_ChJp[df19_ChJp.Q3.isin(["Japan"])]


## Q18(Program_Language): 칼럼번호 83~95 - others
df19_Jp_PL = pd.DataFrame()
df19_Jp_PL['Program_Language'] = [df19_Jp[col][1:].value_counts().index[0] for col in df19_Jp.columns[82:94]]
df19_Jp_PL['counts'] = [df19_Jp[col][1:].value_counts().values[0] for col in df19_Jp.columns[82:94]]


## 2019 China Q18_Part11(None) 결측값 제거
df19_Ch_rmQ18P11 = df19_Ch.drop(['Q18_Part_11'], axis='columns')

## Q18(Program_Language): 칼럼번호 83~95 - others - Q18_Part11(None)
df19_Ch_PL = pd.DataFrame()
df19_Ch_PL['Program_Language'] = [df19_Ch_rmQ18P11[col][1:].value_counts() .index[0] for col in df19_Ch_rmQ18P11.columns[82:93]]
df19_Ch_PL['counts'] = [df19_Ch_rmQ18P11[col][1:].value_counts() .values[0] for col in df19_Ch_rmQ18P11.columns[82:93]]



df19_Jp_PL.insert(0, 'Country',  'Japan')
df19_Ch_PL.insert(0, 'Country',  'China')

df19_PL_JnC = pd.concat([df19_Jp_PL,df19_Ch_PL], ignore_index=True)

In [ ]:
df19_Ch = df19_ChJp[df19_ChJp.Q3.isin(["China"])]
df19_Jp = df19_ChJp[df19_ChJp.Q3.isin(["Japan"])]


## Q18(Program_Language): 칼럼번호 83~95 - others & other(text)
df19_Jp_PL = pd.DataFrame()
df19_Jp_PL['Program_Language'] = [df19_Jp[col][1:].value_counts().index[0] for col in df19_Jp.columns[82:93]]
df19_Jp_PL['counts'] = [df19_Jp[col][1:].value_counts().values[0] for col in df19_Jp.columns[82:93]]


## 2019 China Q18_Part11(None) 결측값 제거
df19_Ch_rmQ18P11 = df19_Ch.drop(['Q18_Part_11'], axis='columns')

## Q18(Program_Language): 칼럼번호 83~95 - others & other(text) - Q18_Part11(None)
df19_Ch_PL = pd.DataFrame()
df19_Ch_PL['Program_Language'] = [df19_Ch_rmQ18P11[col][1:].value_counts() .index[0] for col in df19_Ch_rmQ18P11.columns[82:92]]
df19_Ch_PL['counts'] = [df19_Ch_rmQ18P11[col][1:].value_counts() .values[0] for col in df19_Ch_rmQ18P11.columns[82:92]]



df19_Jp_PL.insert(0, 'Country',  'Japan')
df19_Ch_PL.insert(0, 'Country',  'China')

df19_PL_JnC = pd.concat([df19_Jp_PL,df19_Ch_PL], ignore_index=True)

### Split year{Country}

다른 csv인 2019자료와 2021자료 통합

In [ ]:
df21_PL_JnC.insert(0, 'year',  '2021')
df19_PL_JnC.insert(0, 'year',  '2019')

df_PL_JnC_21n19 = pd.concat([df21_PL_JnC,df19_PL_JnC], ignore_index=True)

Program_Language의 19년도 21년도 통합 value_counts의 정렬(연도 - 언어 - 나라)

values는 Program_Language의 value_counts

color 기준: country

In [ ]:
fig = px.treemap(df_PL_JnC_21n19, path=[px.Constant("2019n2021"),'year','Program_Language','Country'],
                values='counts', color='Country',
                  color_discrete_map={'(?)':'lightgrey', 'China':'gold', 'Japan':'darkblue'})

fig.data[0].textinfo = 'label+percent parent+value'

fig.update_layout(margin = dict(t=50, l=25, r=25, b=25))

fig.show()

## Japan & China: IDE's

Program_Language과정과 동일

### Split Country

#### 2021

In [ ]:
## 앞선 Program_Language에서 선언된 객체로 주석처리
#df21_Ch = df21_ChJp[df21_ChJp.Q3.isin(["China"])]
#df21_Jp = df21_ChJp[df21_ChJp.Q3.isin(["Japan"])]


## Q9(IDE's): 칼럼번호 22~34 - others
df21_Jp_IDEs = pd.DataFrame()
df21_Jp_IDEs['IDE\'s'] = [df21_Jp[col][1:].value_counts().index[0] for col in df21_Jp.columns[21:33]]
df21_Jp_IDEs['counts'] = [df21_Jp[col][1:].value_counts().values[0] for col in df21_Jp.columns[21:33]]


df21_Ch_IDEs = pd.DataFrame()
df21_Ch_IDEs['IDE\'s'] = [df21_Ch[col][1:].value_counts().index[0] for col in df21_Ch.columns[21:33]]
df21_Ch_IDEs['counts'] = [df21_Ch[col][1:].value_counts().values[0] for col in df21_Ch.columns[21:33]]



df21_Ch_IDEs.insert(0, 'Country',  'China')
df21_Jp_IDEs.insert(0, 'Country',  'Japan')

df21_IDEs_JnC = pd.concat([df21_Jp_IDEs,df21_Ch_IDEs], ignore_index=True)

#### 2019

In [ ]:
#df19_Ch = df19_ChJp[df19.Q3.isin(["China"])]
#df19_Jp = df19_ChJp[df19.Q3.isin(["Japan"])]


## Q16(IDE's): 칼럼번호 57~69 - others
df19_Jp_IDEs = pd.DataFrame()
df19_Jp_IDEs['IDE\'s'] = [df19_Jp[col][1:].value_counts().index[0] for col in df19_Jp.columns[56:68]]
df19_Jp_IDEs['counts'] = [df19_Jp[col][1:].value_counts().values[0] for col in df19_Jp.columns[56:68]]


df19_Ch_IDEs = pd.DataFrame()
df19_Ch_IDEs['IDE\'s'] = [df19_Ch[col][1:].value_counts().index[0] for col in df19_Ch.columns[56:68]]
df19_Ch_IDEs['counts'] = [df19_Ch[col][1:].value_counts().values[0] for col in df19_Ch.columns[56:68]]



df19_Jp_IDEs.insert(0, 'Country',  'China')
df19_Ch_IDEs.insert(0, 'Country',  'Japan')

df19_IDEs_JnC = pd.concat([df19_Jp_IDEs,df19_Ch_IDEs], ignore_index=True)

In [ ]:
#df19_Ch = df19_ChJp[df19.Q3.isin(["China"])]
#df19_Jp = df19_ChJp[df19.Q3.isin(["Japan"])]


## Q16(IDE's): 칼럼번호 57~69 - others & other(text)
df19_Jp_IDEs = pd.DataFrame()
df19_Jp_IDEs['IDE\'s'] = [df19_Jp[col][1:].value_counts().index[0] for col in df19_Jp.columns[56:67]]
df19_Jp_IDEs['counts'] = [df19_Jp[col][1:].value_counts().values[0] for col in df19_Jp.columns[56:67]]


df19_Ch_IDEs = pd.DataFrame()
df19_Ch_IDEs['IDE\'s'] = [df19_Ch[col][1:].value_counts().index[0] for col in df19_Ch.columns[56:67]]
df19_Ch_IDEs['counts'] = [df19_Ch[col][1:].value_counts().values[0] for col in df19_Ch.columns[56:67]]



df19_Jp_IDEs.insert(0, 'Country',  'China')
df19_Ch_IDEs.insert(0, 'Country',  'Japan')

df19_IDEs_JnC = pd.concat([df19_Jp_IDEs,df19_Ch_IDEs], ignore_index=True)

### Split year{Country}

In [ ]:
df21_IDEs_JnC.insert(0, 'year',  '2021')
#df21_JCQ9.rename(columns={'Q9':'IDE'}, inplace = True)
df19_IDEs_JnC.insert(0, 'year',  '2019')
#df19_JCQ9.rename(columns={'Q16':'IDE'}, inplace = True)

df_IDEs_JnC_21n19 = pd.concat([df21_IDEs_JnC,df19_IDEs_JnC], ignore_index=True)

IDE's 요소명 간랸화

[pandas replace](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.replace.html)

In [ ]:
#df_IDEs_JnC_21n19.replace('Jupyter (JupyterLab, Jupyter Notebooks, etc)', 'Jupyter', inplace = True)
#df_IDEs_JnC_21n19.replace('Visual Studio / Visual Studio Code', 'VS / VSCode', inplace = True)
#df_IDEs_JnC_21n19.replace(regex = 'Jupyter (JupyterLab, Jupyter Notebooks, etc) ', value = 'Jupyter', inplace = True)
df_IDEs_JnC_21n19.replace(to_replace = 'Jupyter (JupyterLab, Jupyter Notebooks, etc) ', value = 'Jupyter', inplace = True)
df_IDEs_JnC_21n19.replace(to_replace = 'Visual Studio / Visual Studio Code', value = 'VS / VSCode', inplace = True, regex = True)

In [ ]:
fig = px.treemap(df_IDEs_JnC_21n19, path=[px.Constant("2019n2021"),'year','IDE\'s','Country'],
                values='counts', color='Country',
                  color_discrete_map={'(?)':'lightgrey', 'China':'gold', 'Japan':'darkblue'})

fig.data[0].textinfo = 'label+percent parent+value'

fig.update_layout(margin = dict(t=50, l=25, r=25, b=25))

fig.show()

### 외부링크

 - [데이터프레임 합치기](https://yganalyst.github.io/data_handling/Pd_12/)
 - [행열 이름 변경](https://mizykk.tistory.com/131)
 - [pd insert](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.insert.html)
     - [Ex1](https://www.delftstack.com/ko/howto/python-pandas/how-to-add-new-column-to-existing-dataframe-in-python-pandas/)
     - [Ex2](https://m.blog.naver.com/PostView.naver?isHttpsRedirect=true&blogId=rising_n_falling&logNo=221631637822)

## ML(machine learning) frameworks

In [ ]:
df21_ChJp = df21[df21.Q3.isin(["Japan","China"])]

df21_ChJp_total_ML = pd.DataFrame()
df21_ChJp_total_ML['ML'] = [df21_ChJp[col][1:].value_counts().index[0] for col in df21_ChJp.columns[72:91]]
df21_ChJp_total_ML['counts'] = [df21_ChJp[col][1:].value_counts().values[0] for col in df21_ChJp.columns[72:91]]

In [ ]:
fig = px.treemap(df21_ChJp_total_ML, path=[px.Constant("2021: Jp+Ch"),'ML'],
                values='counts', color='counts')

fig.data[0].textinfo = 'label+percent parent'

fig.show()

In [ ]:
df19_ChJp = df19[df19.Q3.isin(["Japan","China"])]

df19_ChJp_total_ML = pd.DataFrame()
df19_ChJp_total_ML['ML'] = [df19_ChJp[col][1:].value_counts().index[0] for col in df19_ChJp.columns[155:168]]
df19_ChJp_total_ML['counts'] = [df19_ChJp[col][1:].value_counts().values[0] for col in df19_ChJp.columns[155:168]]

In [ ]:
fig = px.treemap(df19_ChJp_total_ML, path=[px.Constant("2019: Jp+Ch"),'ML'],
                values='counts', color='counts')

fig.data[0].textinfo = 'label+percent parent'

fig.show()